courses/utils.py

```diff
@@ -25,15 +26,26 @@ def ensure_course_run_grade(user, course_run, edx_grade, should_update=False):
 
     updated = False
     if should_update:
-        grade, created = CourseRunGrade.objects.update_or_create(
-            course_run=course_run, user=user, defaults=grade_properties
-        )
-        updated = not created
+        with transaction.atomic():
+            run_grade, created = CourseRunGrade.objects.select_for_update().get_or_create(
+                course_run=course_run,
+                user=user,
+                defaults=grade_properties
+            )
+
+            if not created and not run_grade.set_by_admin and run_grade.grade != edx_grade.percent:
+                # Perform actual update now.
+                run_grade.grade = edx_grade.percent
+                run_grade.passed = edx_grade.passed
+                run_grade.letter_grade = edx_grade.letter_grade
+                run_grade.save_and_log(None)
+                updated = not created
```

Comment:
`updated = True` as you're already enforcing the `if not created` on this block. Or you could move this line above the condition.